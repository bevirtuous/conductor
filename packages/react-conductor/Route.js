module.exports = require('./src/Route');
