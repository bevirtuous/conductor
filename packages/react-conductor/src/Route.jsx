import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Transition from 'react-transition-group/Transition';
import uuid from 'uuid/v4';
import Conductor from '@virtuous/conductor';

/**
 * The Route component.
 */
class Route extends Component {
  static propTypes = {
    content: PropTypes.func.isRequired,
    pattern: PropTypes.string.isRequired,
    component: PropTypes.string,
    index: PropTypes.number,
    isOpen: PropTypes.bool,
    isVisible: PropTypes.bool,
    path: PropTypes.string,
    transition: PropTypes.shape(),
  };

  static defaultProps = {
    component: 'div',
    index: null,
    isOpen: false,
    isVisible: false,
    path: null,
    transition: null,
  };

  static childContextTypes = {
    route: PropTypes.shape(),
  };

  /**
   * The constructor.
   * @param {Object} props The component props.
   */
  constructor(props) {
    super(props);

    // Generate a unique ID for this route.
    this.id = uuid();

    // Register this route with the router.
    Conductor.register(this.id, this.props.pattern);
  }

  /**
   * Creates a context about this route.
   * @returns {Object}
   */
  getChildContext() {
    return {
      route: {
        id: this.id,
      },
    };
  }

  /**
   * @param {Object} nextProps The next set of props.
   * @returns {boolean}
   */
  shouldComponentUpdate(nextProps) {
    return (
      this.props.isOpen !== nextProps.isOpen ||
      this.props.isVisible !== nextProps.isVisible
    );
  }

  /**
   * Fire the enter callback when this is the new active route.
   */
  componentDidUpdate() {
    const currentRoute = Conductor.getCurrentRoute();
    if (!this.props.transition && this.props.path === currentRoute.pathname) {
      this.routeDidEnter();
    }
  }

  /**
   * Returns a transition object based on the state of the route.
   * @returns {Object}
   */
  get transitionType() {
    const currentAction = Conductor.getCurrentAction();
    
    if (this.props.isOpen) {
      return this.props.transition.backward;
    } else if (currentAction === 'REPLACE') {
      return this.props.transition.replace;
    }

    return this.props.transition.forward;
  }

  /**
   * Inform the Conductor that the route has entered.
   */
  routeDidEnter = () => {
    const currentAction = Conductor.getCurrentAction();

    switch (currentAction) {
      case 'POP':
        Conductor.popped(this.id, this.props.path);
        break;
      case 'REPLACE':
        Conductor.replaced(this.id, this.props.path);
        break;
      default:
        Conductor.pushed(this.id, this.props.path);
        break;
    }
  }

  /**
   * Renders the component.
   * @returns {JSX}
   */
  render() {
    const {
      component: Wrapper,
      content: Content,
      index,
      isVisible,
      transition,
    } = this.props;

    if (!transition) {
      return (
        <Wrapper
          style={{
            ...(index !== null) && { zIndex: index },
            ...(!isVisible) && { display: 'none' },
          }}
        >
          <Content />
        </Wrapper>
      );
    }

    /**
     * Switch between the transitions based on whether or not
     * this route is open or if the route was replaced.
     */
    const { transitionType } = this;

    return (
      <Transition
        in={isVisible}
        timeout={transitionType.duration}
        onEntered={this.routeDidEnter}
      >
        {state => (
          <Wrapper
            style={{
              ...(index !== null) && { zIndex: index },
              ...transitionType.default,
              ...transitionType[state],
            }}
          >
            <Content />
          </Wrapper>
        )}
      </Transition>
    );
  }
}

export default Route;
