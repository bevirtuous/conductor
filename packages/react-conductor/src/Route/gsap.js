const forward = {
  duration: 375,
  default: {
    height: '100vh',
    left: '0',
    position: 'absolute',
    top: '0',
    transform: 'translate3d(100%, 0, 0)',
    width: '100%',
    willChange: 'transform',
  },
  entering: {
    pointerEvents: 'none',
    transform: 'none',
  },
  entered: {
    transform: 'none',
  },
  exiting: {
    pointerEvents: 'none',
    transform: 'translate3d(100%, 0, 0)',
  },
  exited: {
    pointerEvents: 'none',
    transform: 'translate3d(100%, 0, 0)',
  },
};

const backward = {
  duration: 375,
  default: {
    height: '100vh',
    left: '0',
    position: 'absolute',
    top: '0',
    transform: 'translate3d(-50%, 0, 0)',
    width: '100%',
    willChange: 'transform',
  },
  entering: {
    pointerEvents: 'none',
    transform: 'none',
  },
  entered: {
    transform: 'none',
  },
  exiting: {
    pointerEvents: 'none',
    transform: 'translate3d(-50%, 0, 0)',
  },
  exited: {
    pointerEvents: 'none',
    transform: 'translate3d(-50%, 0, 0)',
  },
};

const replace = {
  duration: 375,
  default: {
    height: '100vh',
    left: '0',
    position: 'absolute',
    top: '0',
    transform: 'none',
    transition: 'opacity 375ms cubic-bezier(0.25, 0.1, 0.25, 1), transform 375ms cubic-bezier(0.25, 0.1, 0.25, 1)',
    width: '100%',
    willChange: 'transform',
  },
  entering: {
    pointerEvents: 'none',
    transform: 'none',
  },
  entered: {
    transform: 'none',
  },
  exiting: {
    opacity: '0.2',
    pointerEvents: 'none',
    transform: 'translate3d(0, -10%, 0)',
  },
  exited: {
    opacity: 0,
    pointerEvents: 'none',
    transform: 'translate3d(0, -10%, 0)',
  },
};

export default {
  forward,
  backward,
  replace,
};
