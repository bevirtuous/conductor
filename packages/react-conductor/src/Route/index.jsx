import React from 'react';
import PropTypes from 'prop-types';
import { router } from '@virtuous/conductor';
import { RouteContext, RouterContext } from '../context';

/**
 * The Route component.
 */
class Route extends React.Component {
  static contextType = RouterContext;

  static propTypes = {
    component: PropTypes.func.isRequired,
    pattern: PropTypes.string.isRequired,
    transform: PropTypes.func,
  }

  static defaultProps = {
    transform: null,
  }

  /**
   * @param {Object} props The component props.
   */
  constructor(props) {
    super(props);
    router.register(props.pattern, props.transform);
  }

  /**
   * TODO: Move to router
   */
  get currentRoute() {
    const { [router.routeIndex]: [, route] } = this.context;

    return route;
  }

  /**
   * @returns {Object}
   */
  get contextValue() {
    const { setPattern, ...context } = this.currentRoute;

    return context;
  }

  /**
   * @returns {JSX}
   */
  render() {
    const { component: Component, pattern } = this.props;
    const route = this.currentRoute;

    if (route.pattern !== pattern) {
      return null;
    }

    return (
      <RouteContext.Provider key={route.id} value={this.contextValue}>
        <Component />
      </RouteContext.Provider>
    );
  }
}

export default Route;
