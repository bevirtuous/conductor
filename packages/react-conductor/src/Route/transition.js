const transition = {
  default: {
    height: '100vh',
    left: '0',
    position: 'absolute',
    top: '0',
    width: '100%',
  },
  exiting: {
    display: 'none',
  },
  exited: {
    display: 'none',
  },
};

export default {
  forward: transition,
  backward: transition,
  replace: transition,
};
