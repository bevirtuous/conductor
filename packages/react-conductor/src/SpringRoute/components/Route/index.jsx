import React from 'react';
import PropTypes from 'prop-types';
import { Spring } from 'react-spring';
import { emitter, router } from '@virtuous/conductor';
import { RouteContext } from '../../../context';
import * as constants from '../../../constants';

/**
 * The SpringRoute Component.
 */
class SpringRoute extends React.Component {
  static contextType = RouteContext;

  static propTypes = {
    component: PropTypes.func.isRequired,
    current: PropTypes.bool.isRequired,
    index: PropTypes.number.isRequired,
    transition: PropTypes.shape().isRequired,
    className: PropTypes.string,
  }

  static defaultProps = {
    className: null,
  }

  state = {
    render: true,
  }

  /**
   * @returns {Object}
   */
  get transition() {
    const { current, transition } = this.props;
    const subset = transition[router.action.toLowerCase()];

    // TODO: check for transition passed through route action.

    if (!current) {
      return subset.out;
    }

    return subset.in;
  }

  /**
   * When the route is the current route, make sure that it can render.
   * @param {Object} props The new component props.
   * @returns {Object|null}
   */
  static getDerivedStateFromProps(props) {
    if (props.current) {
      return { render: true };
    }

    return null;
  }

  handleStart = () => {
    const { current } = this.props;
    const eventName = current ? constants.EVENT_WILL_ENTER : constants.EVENT_WILL_LEAVE;

    emitter.emit(eventName, this.context);
  }

  /**
   * When the animation has completed, we need to set the render state of the component.
   * If the component animated out then it should not be rendered.
   */
  handleRest = () => {
    const { current } = this.props;

    if (!current) {
      this.setState(
        { render: true },
        () => emitter.emit(constants.EVENT_DID_LEAVE, this.context)
      );
    } else {
      emitter.emit(constants.EVENT_DID_ENTER, this.context);
    }
  }

  /**
   * @returns {JSX|null}
   */
  render() {
    const { render } = this.state;

    if (!render) {
      return null;
    }

    const { className, component: Component, index } = this.props;

    return (
      <Spring {...this.transition} onStart={this.handleStart} onRest={this.handleRest}>
        {props => (
          <div className={className} style={{ ...props, zIndex: index }}>
            <Component />
          </div>
        )}
      </Spring>
    );
  }
}

export default SpringRoute;
