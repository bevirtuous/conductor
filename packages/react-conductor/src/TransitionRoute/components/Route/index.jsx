import React from 'react';
import { Spring } from 'react-spring';

// How to know when a route has to leave?
  // When the stack changes and the route is no longer at current index (previous routes only)
// How to know when a route is intended to be gone? (forward routes)
  // Not in stack at all
  // How to transition out?
  // Stateful component!

  // internally the route can check to see if it should render it's content or not
  // (if it's a next route and has animated out)

/**
 * The SpringRoute Component.
 */
class SpringRoute extends React.Component {
  /**
   * 
   * @param {*} props 
   */
  constructor(props) {
    super(props);

    this.state = {
      in: false,
      render: true,
    };
  }

  componentWillReceiveProps() {
    // Check if the route became visible
    // Check if 
    // Case 1. Route leaves via push (in history)
      // How to know how to move?

    // Case 2. Route leaves via pop (removed from history)
      // How to know how to move?
      // Update render state after animating
  }

  /**
   * @returns {Object}
   */
  get transition() {
    const { component: Component } = this.props;
    return {
      from: { x: window.innerWidth },
      to: { x: 0 },
    };
  }

  /**
   * @returns {JSX|null}
   */
  render() {
    const { render } = this.state;

    if (!render) {
      return null;
    }

    const { component: Component } = this.props;
    return (
      <Spring {...this.transition}>
        {props => (
          <div style={{ transform: `translateX(${props.x}px)` }}>
            <Component />
          </div>
        )}
      </Spring>
    );
  }
}

export default SpringRoute;
