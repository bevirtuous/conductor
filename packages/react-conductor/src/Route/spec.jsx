import React from 'react';
import { mount } from 'enzyme';
import { router } from '@virtuous/conductor';
import Router from '../Router/index';
import { RouteContext } from '../context';
import Route from './index';

const pattern = '/myroute/:id';
const spy = jest.spyOn(router, 'register');
const transform = route => route;

describe('<Route />', () => {
  it('should render component', () => {
    mount((
      <Router>
        <Route
          pattern={pattern}
          component={() => <div />}
          transform={transform}
        />
      </Router>
    ));

    expect(spy).toHaveBeenCalledWith(pattern, transform);
  });

  it('should correctly set the RouteContext value', () => {
    let contextValue = null;
    const current = router.getCurrentRoute();

    const MyComponent = () => (
      <RouteContext.Consumer>
        {(route) => {
          contextValue = route;
          return null;
        }}
      </RouteContext.Consumer>
    );

    mount((
      <Router>
        <Route
          pattern={pattern}
          component={MyComponent}
        />
      </Router>
    ));

    expect(current).toMatchObject(contextValue);
  });

  it('should render null when does not match current route', () => {
    const route = mount((
      <Router>
        <Route
          pattern="/wrong"
          component={() => <div />}
        />
      </Router>
    ));

    expect(route.html()).toBe('');
  });

  it('should react to router events and update', async () => {
    const route = mount((
      <Router>
        <Route
          pattern={pattern}
          component={() => <div />}
        />
        <Route
          pattern="/other"
          component={() => <div />}
        />
      </Router>
    ));

    // Should initially render the component (match found).
    expect(route.find(Route).at(0).html()).toBe('<div></div>');
    expect(route.find(Route).at(1).html()).toBeNull();

    await router.push({ pathname: '/other' });

    route.update();

    // Should now render null (first route does not match).
    expect(route.find(Route).at(0).html()).toBeNull();
    expect(route.find(Route).at(1).html()).toBe('<div></div>');
  });
});
