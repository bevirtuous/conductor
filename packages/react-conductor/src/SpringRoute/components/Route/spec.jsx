import React from 'react';
import { mount } from 'enzyme';
import Route from '.';

describe('SpringRoute Child', () => {
  it('should render as expected', () => {
    const app = mount((
      <Route
        className="my-classname"
        component={() => <div />}
        current
        index={1}
        spring={() => {}}
      />
    ));

    expect(app).toMatchSnapshot();
  });

  it('should not render when `render` state is false', (done) => {
    const app = mount((
      <Route
        className="my-classname"
        component={() => <div />}
        current={false}
        index={1}
        spring={() => {}}
      />
    ));

    app.setState({ render: false }, () => {
      expect(app).toMatchSnapshot();
      done();
    });
  });
});
