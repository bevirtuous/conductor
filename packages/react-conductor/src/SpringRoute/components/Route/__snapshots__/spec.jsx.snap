// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpringRoute Child should not render when \`render\` state is false 1`] = `
<SpringRoute
  className="my-classname"
  component={[Function]}
  current={false}
  index={1}
  spring={[Function]}
/>
`;

exports[`SpringRoute Child should render as expected 1`] = `
<SpringRoute
  className="my-classname"
  component={[Function]}
  current={true}
  index={1}
  spring={[Function]}
>
  <Spring
    config={
      Object {
        "friction": 26,
        "tension": 170,
      }
    }
    force={false}
    from={Object {}}
    immediate={true}
    inject={[Function]}
    native={false}
    onRest={[Function]}
    onStart={[Function]}
    reset={false}
    to={Object {}}
  />
</SpringRoute>
`;
